### PR TITLE
Light Mode: implement getPoolLifeCycleStatus 

### PR DESCRIPTION
**Pull request by Unisay**
I have implemented `getPoolLifeCycleStatus` function of the `StakePoolLayer` for the light mode.

Tested in REPL on Testnet:
```
Haskell λ cfg >>= \c -> _getPoolLifeCycleStatus c net pid
PoolRegistered (PoolRegistrationCertificate {poolId = PoolId {getPoolId = "]\153(+\187H@8\v\185\140\aT\152\237\EM\131\174\225\138J\t%\185\180M\147\241"}, poolOwners = [PoolOwner {getPoolOwner = "LT\247\187\226iP\239\221\250+\249\246\159\247\254\243\132G\140G*\209h\EMv\169\DC3"}], poolMargin = Percentage (7782220156096217 % 288230376151711744), poolCost = Coin 340000000, poolPledge = Coin 100000000, poolMetadata = Just (StakePoolMetadataUrl "https://git.io/JJ7wm",StakePoolMetadataHash "M\137\ENQL)b!V\148\167\DC2-\254A\188r\141>\194H\248\SO\169\162\224\212\147\ENQ}}#8")})
```

### Issue Number

ADP-1511

### CODE REVIEW

File: lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost/Conversion.hs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( fromText )
+
+import qualified Blockfrost.Client as BF
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address )
+
+fromBfLovelaces :: MonadError BlockfrostError m => BF.Lovelaces -> m Coin
```
Comment (HeinrichApfelmus) [inline lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost/Conversion.hs:38]: I'm still not too fond of the faux generality of `MonadError` and would prefer `Either BlockfrostError Coin` as a result (at the cost of a `lift*`) — but at least it does seem that the explicit return types provide a smoother experience than the `FromBlockfrost` type class. So 👍 on explicit conversion functions.

Comment (Unisay) [reply lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost/Conversion.hs:38]: Yeah, ok. Let's say that first step is to get rid of the typeclass in favor of functions (`Conversion` module). Then the second step could address Either/lifting.

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 2]
```diff
@@ -1,10 +1,8 @@
-{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
-
+{-# LANGUAGE BlockArguments #-}
```
Comment (HeinrichApfelmus) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:1]: (I'm still not foo fond of it, as it messes up the "whitespace separates function arguments" property. 😅)

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 3]
```diff
@@ -218,15 +263,150 @@ data StakePoolLayer = StakePoolLayer
...
+        }
+
+_getPoolLifeCycleStatus
+    :: BF.ClientConfig
+    -> SomeNetworkDiscriminant
+    -> PoolId
+    -> IO PoolLifeCycleStatus
+_getPoolLifeCycleStatus
+    cfg network@(SomeNetworkDiscriminant (Proxy :: Proxy nd)) poolId =
+        BFM.run cfg do
```
Comment (HeinrichApfelmus) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:327]:
```suggestion
      BFM.run cfg do
```
Style: Clearer indentation for following `do` block?

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 4]
```diff
@@ -218,15 +263,150 @@ data StakePoolLayer = StakePoolLayer
...
+        let bfPoolId = BF.PoolId (toText poolId)
+        poolUpdates <- BF.allPages \paged ->
+            BF.getPoolUpdates' bfPoolId paged BF.Ascending
+        case reverse poolUpdates of
```
Comment (HeinrichApfelmus) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]:
```suggestion
            BF.getPoolUpdates' bfPoolId paged BF.Descending
        case poolUpdates of
```
Possible? Not sure if preferable, though. 🤔

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 5]
```diff
@@ -218,15 +263,150 @@ data StakePoolLayer = StakePoolLayer
...
+                    BF.PoolDeregistered ->
+                        case findRegCert otherUpdates of
+                            Just regCertUpdate -> do
+                                reg <- poolRegCert regCertUpdate
+                                drg <- poolDeregCert lastUpdate
+                                pure $ PoolRegisteredAndRetired reg drg
+                            Nothing ->
+                                throwError $ PoolRegistrationIsMissing poolId
+
+    where
```
Comment (HeinrichApfelmus) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]:
```suggestion
  where
```
Style: 2 spaces for `where` for better visual distinction.

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 6]
```diff
@@ -733,27 +905,25 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
...
+            FetchSMASH (unSmashServer -> uri) -> do
+                STM.atomically $ writeTVar gcStatus NotStarted
+                let getDelistedPools =
+                        fetchDelistedPools trFetch uri manager
+                tid <- forkFinally
+                    (gcDelistedPools gcStatus tr db getDelistedPools)
+                    (traceAfterThread (contramap MsgGCThreadExit tr))
+                void $ fetchMetadata manager [registryUrlBuilder uri]
+                    `finally` killThread tid) else
+        traceWith tr MsgSMASHUnreachable
```
Comment (HeinrichApfelmus) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]:
```suggestion
    if health == Available || health == NoSmashConfigured
    then do
        case poolMetadataSource settings of
            FetchNone -> do
                STM.atomically $ writeTVar gcStatus NotApplicable

            FetchDirect -> do
                STM.atomically $ writeTVar gcStatus NotApplicable
                void $ fetchMetadata manager [identityUrlBuilder]

            FetchSMASH (unSmashServer -> uri) -> do
                STM.atomically $ writeTVar gcStatus NotStarted
                let getDelistedPools =
                        fetchDelistedPools trFetch uri manager
                tid <- forkFinally
                    (gcDelistedPools gcStatus tr db getDelistedPools)
                    (traceAfterThread (contramap MsgGCThreadExit tr))
                void $ fetchMetadata manager [registryUrlBuilder uri]
                    `finally` killThread tid
    else
        traceWith tr MsgSMASHUnreachable
```

Style: Good old `if-then-else` without parentheses. 😄

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 7]
```diff
@@ -769,16 +939,16 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
...
-                            settings' <- readSettings
-                            when (settings == settings') $ putPoolMetadata hash meta
+                Nothing ->
+                    atomically $ do
+                        settings' <- readSettings
+                        when (settings == settings') $ putFetchAttempt (url, hash)
+                Just meta -> do
+                    atomically $ do
+                        settings' <- readSettings
+                        when (settings == settings') $ putPoolMetadata hash meta
```
Comment (HeinrichApfelmus) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]:
```suggestion
                    Nothing ->
                        atomically $ do
                            settings' <- readSettings
                            when (settings == settings') $ putFetchAttempt (url, hash)
                    Just meta -> do
                        atomically $ do
                            settings' <- readSettings
                            when (settings == settings') $ putPoolMetadata hash meta
```

Style: I would prefer to keep the old indentation, as the `case` expression would be indented also in the event where it had come from a variable, i.e.

```
foo <- fetchFromRemote trFetch strategies manager pid url hash
case foo of
    Nothing -> …

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 8]
```diff
@@ -236,54 +416,46 @@ newStakePoolLayer gcStatus nl db@DBLayer {..} restartSyncThread = do
...
-    _getSettings :: IO Settings
-    _getSettings = liftIO $ atomically readSettings
-
     _listPools
         :: EpochNo
         -- Exclude all pools that retired in or before this epoch.
         -> Coin
         -> IO [Api.ApiStakePool]
     _listPools currentEpoch userStake = do
-        rawLsqData <- liftIO $ stakeDistribution nl userStake
```
Comment (HeinrichApfelmus) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:276]: Not sure where the `liftIO` came from, thanks for removing it! 😊

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 9]
```diff
@@ -218,15 +263,150 @@ data StakePoolLayer = StakePoolLayer
...
+                                for _poolUpdateMetadataHash
+                                    stakePoolMetadataHashFromText
+                            pure (url, hash)
+                pure PoolRegistrationCertificate{..}
+            Nothing ->
+                throwError $ PoolUpdateCertificateNotFound poolId update
+
+    poolDeregCert :: BF.PoolUpdate -> BFM PoolRetirementCertificate
+    poolDeregCert update@BF.PoolUpdate{..} = do
+        poolRetiring  <-
```
Comment (HeinrichApfelmus) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]:
```suggestion
        poolRetiring <-
```

**Review by HeinrichApfelmus — APPROVED** [review]
Looks good to me! 😊 Small style comments. Thanks for removing the unnecessary `liftIO`.

File: lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost/Conversion.hs (hunk 1 above)
Comment (HeinrichApfelmus) [reply lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost/Conversion.hs:38]: Sounds good! 🤝

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs (hunk 3 above)
Comment (Unisay) [reply lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:327]: Ok

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs (hunk 7 above)
Comment (Unisay) [reply lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: 👌🏻

**Pull request merged**